### PR TITLE
Convert and re-enable TimingModuleTest

### PR DESCRIPTION
**Pull request by amrfarid140**
## Summary:

As part of the effort to Kotlin-fy React Native tests, I've converted [TimingModuleTest](https://github.com/facebook/react-native/blob/main/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/modules/timing/TimingModuleTest.java) to Kotlin.

I noticed the test suite was ignored and it was failing locally but managed to get all the tests to pass except one which I kept ignored.

## Changelog:

[Internal] [Changed] - Convert TimingModuleTest to Kotlin

## Test Plan:

Tests pass: ./gradlew :packages:react-native:ReactAndroid:test
Formatted with [KtFmt](https://facebook.github.io/ktfmt/)

### CODE REVIEW

File: packages/react-native/ReactAndroid/src/test/java/com/facebook/react/modules/timing/TimingModuleTest.kt [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+    assertThat(mTimingModule.hasActiveTimersInRange(200)).isTrue // In range
+  }
+
+  @Test
+  @Ignore("Failing due to failed invocation verification") // TODO T13905097
+  fun testIdleCallback() {
+    mTimingModule.onHostResume()
+    mTimingModule.setSendIdleEvents(true)
+    stepChoreographerFrame()
+    verify(mJSTimersMock).callIdleCallbacks(SystemClock.currentTimeMillis().toDouble())
```
Comment (amrfarid140) [inline packages/react-native/ReactAndroid/src/test/java/com/facebook/react/modules/timing/TimingModuleTest.kt:233]: This verification fails for some reason. I've decided that try to fix it is out of scope for this PR so will attempt to fix it in a subsequent PR.

File: packages/react-native/ReactAndroid/src/test/java/com/facebook/react/modules/timing/TimingModuleTest.kt [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+@[
+PrepareForTest(Arguments::class, SystemClock::class, ReactChoreographer::class)
+PowerMockIgnore("org.mockito.*", "org.robolectric.*", "androidx.*", "android.*")
+RunWith(RobolectricTestRunner::class)]
+class TimingModuleTest {
+  companion object {
+    const val FRAME_TIME_NS = 17 * 1000 * 1000
+  }
+
+  private lateinit var mTimingModule: TimingModule
```
Comment (rshest) [inline packages/react-native/ReactAndroid/src/test/java/com/facebook/react/modules/timing/TimingModuleTest.kt:45]: Please avoid using Hungarian notation in Kotlin, `timingModule` is a better name in this case (ditto for all the others).

File: packages/react-native/ReactAndroid/src/test/java/com/facebook/react/modules/timing/TimingModuleTest.kt (hunk 1 above)
Comment (rshest) [reply packages/react-native/ReactAndroid/src/test/java/com/facebook/react/modules/timing/TimingModuleTest.kt:233]: Did it fail previously, when it was in Java?

File: packages/react-native/ReactAndroid/src/test/java/com/facebook/react/modules/timing/TimingModuleTest.kt [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    val reactInstance = mock(CatalystInstance::class.java)
+    val reactContext = mock(ReactApplicationContext::class.java)
+    mockWhen(reactContext.catalystInstance).thenReturn(reactInstance)
+    mockWhen(reactContext.hasActiveReactInstance()).thenReturn(true)
+
+    mPostFrameCallbackHandler = PostFrameCallbackHandler()
+    mIdlePostFrameCallbackHandler = PostFrameIdleCallbackHandler()
+
+    doAnswer(mPostFrameCallbackHandler)
+      .`when`(mReactChoreographerMock)
```
Comment (cortinico) [inline packages/react-native/ReactAndroid/src/test/java/com/facebook/react/modules/timing/TimingModuleTest.kt:85]: can you static import this `when` as well?

File: packages/react-native/ReactAndroid/src/test/java/com/facebook/react/modules/timing/TimingModuleTest.kt (hunk 1 above)
Comment (cortinico) [reply packages/react-native/ReactAndroid/src/test/java/com/facebook/react/modules/timing/TimingModuleTest.kt:233]: @rshest yes the whole test suite was suppressed when in Java so this is net positive.

File: packages/react-native/ReactAndroid/src/test/java/com/facebook/react/modules/timing/TimingModuleTest.kt [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    }
+
+    fun getAndResetFrameCallback(): FrameCallback? {
+      val callback = mFrameCallback
+      mFrameCallback = null
+      return callback
+    }
+  }
+
+  private class PostFrameIdleCallbackHandler : Answer<Unit> {
```
Comment (cortinico) [inline packages/react-native/ReactAndroid/src/test/java/com/facebook/react/modules/timing/TimingModuleTest.kt:251]: Those two classes are the same. Can we delete one of the two and make sure we instantiate two of them instead

**Review by cortinico — CHANGES_REQUESTED** [review]
Thanks for taking a stab at this @amrfarid140
Also thank you very much for going the extra mile and re-enabling some of the broken tests!

I've left some comments that needs to be addressed before this can be merged

File: packages/react-native/ReactAndroid/src/test/java/com/facebook/react/modules/timing/TimingModuleTest.kt (hunk 3 above)
Comment (amrfarid140) [reply packages/react-native/ReactAndroid/src/test/java/com/facebook/react/modules/timing/TimingModuleTest.kt]: It is not an import. It's an interface function (see `Stubber` type which is returned by `doAnswer) so we are stuck with the naming. 

But I think we can work around that by starting with `whenever`... will test it out.

**Review by cortinico — APPROVED** [review]
LGTM 👍 thank you again

File: packages/react-native/ReactAndroid/src/test/java/com/facebook/react/modules/timing/TimingModuleTest.kt [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+    timingModule.createTimer(42.0, 150.0, 0.0, false)
+    assertThat(timingModule.hasActiveTimersInRange(100)).isFalse // Out of range
+    assertThat(timingModule.hasActiveTimersInRange(200)).isTrue // In range
+  }
+
+  @Test
+  @Ignore("Failing due to failed invocation verification") // TODO T13905097
+  fun testIdleCallback() {
+    timingModule.onHostResume()
+    timingModule.setSendIdleEvents(true)
```
Comment (amrfarid140) [inline packages/react-native/ReactAndroid/src/test/java/com/facebook/react/modules/timing/TimingModuleTest.kt:238]: @cortinico Just realised this test will pass if we swap these two statements. Should we include it in this PR?

**Comment by amrfarid140** [comment]
@cortinico FYI I've fixed the ignored test. Turns out its a simple case of ordering calls as the implementation expect.

When we call `onHostResume`.. it goes into this flow 

```
  private void maybeSetChoreographerIdleCallback() {
    synchronized (mIdleCallbackGuard) {
      if (mSendIdleEvents) {
        setChoreographerIdleCallback();
      }
    }
  }
```

So if we don't set idle events before it then no callback will be set which means nothing is triggered for idle callbacks which results in the test failing.

**Comment by rshest** [comment]
@amrfarid140 That's awesome, thank you so much!

**Pull request closed without merging**